### ISSUE TITLE
Consolidate test images and improve documentation

### ISSUE THREAD

**Issue by johnSchnake** [kubernetes/kubernetes#76342]
**What would you like to be added**:
I would like the number of test images to be reduced and the documentation improved. The list is in this file:
https://github.com/kubernetes/kubernetes/blob/master/test/utils/image/manifest.go

Looking at this list, it is not clear why we have so many images or why/how that image is being used.

In many cases tests just want to confirm image=`X` and/or after an update the image changes to `Y` and other types of non-functional tests.

The ways to reduce the image count include:
 - In all the places that the image doesn't matter, ensure we are using the same dummy images.
 - In cases where there is very little needed functionality we may be able to just make a generic test image which provides all the needed functionality. In this way we could potentially aggregate 10 images worth of functionality into 1 image.

We still need the images to support multi-arch and use manifests.

**Why is this needed**:
 - It makes it easier to understand what test images we have and why we have them (for any dev that looks at it)
 - It makes it easier to understand which image you should choose if you are writing a new test and need some guidance
 - It reduces the number/size of the images needed in order to test the cluster which is important for testing air-gapped clusters.

Currently there are 47 images, which end up at 7.4 GB

**Comment by johnSchnake** [comment]
Findings so far:
 - 2 images (Cuda images for GPU/vector stuff) take up 4GB of the total 7. This obviously makes it clear that the best approach wont necessarily be to focus on combining 20 other golang images to save 50MB. Instead, perhaps focus should be on splitting up the "master map" of images into focused sections so that it is more clear what each testing set needs?
    - there are 2 versions of the same test and actually no tests that use the newer version yet. This seems a waste as someone outside that domain, but we don't want to be overly strict about that. But we shouldn't confuse people into thinking they might actually need those 2GB files if they dont.
 - Conformance tests need, if I've done this check correctly, only about 14 images (I didnt double check the sizes of these 14 but if each is ~50MB on average that would only be 0.7GB for conformance tests)
 - Documentation was sporadic. On the images that we actually build there was a README (random guess) 40% of the time. I would suggest as part of this effort we strive to make it 100%. If we are custom building and maintaining these images it makes sense to at least say what it does/why it was needed. Kudos so those authors/maintainers that did that; there were some images with really good docs.

FWIW, to get some of this info without a manual slog through every test I did the following:
I tried running a check by making a custom build of the conformance suite which logged when someone called the `imageutils.GetImage` method to see what image they were calling. The one big uncertainty here is that _lots_ of calls to this happen as part of `init()` and I threw those out for this check. So 14 is a _lower_ bound, depending on how many of those init calls were actually later used in a test.

In an ideal world, I could do a sort of `dryRun` which would still execute my method so that I could iterate quickly. In my case, I actually had to wait for the conformance suite to finish in order to get results. If someone has a tip; it would be appreciated.

**Comment by johnSchnake** [comment]
Update: Not much extra work on this but I am of the mind that we split up the list of images due to the findings above. I'm not talking about decentralizing management, but simply providing some separation so that you can clearly see which test images are needed for conformance vs everything else.

This would mean people running airgap conformance testing would need < 1G of images today and, if we do more deduping work there, maybe significantly less. Ideally, we could shoot for one conformance test image but I don't see that being completely achievable or maintainable (e.g. would it be a requirement that a test being upgraded to a Conformance test use that image? That just introduces extra test & image changes which slow down progress there).

**Comment by mgdevstack** [comment]
To sum-up discussion held in another PR, we need to address following challenges: -
1. We emphasise on using 1 container per process but as here discussed in https://github.com/kubernetes/kubernetes/pull/76443#issuecomment-492496994, we are thinking to consolidate multiple images into single cli based app container and installing more binaries like `curl`, `netstat` ..., Either we need to manage multi ENTRYPOINTS or need to come up with `agnhost exec` kind of app which would allow to run any system level binaries. (similar to `kubectl exec` or `docker exec`)
e.g. ``` agnhost exec -- netstat -lt``` 

2. https://github.com/kubernetes/kubernetes/pull/76443#issuecomment-494176802
>Regarding ss, we'll probably have to go with netstat, although the output and the parameters are different on both OSes. I wonder if we could find a Linux-like netstat binary for Windows which we can include, otherwise we'd have to take care of the differences in the agnhost image.

Any good way to address this? because due to this we may end up having `agnhost:linux` and `agnhost:windows` kind of 2 images (harder to manage same functionality on both image, platform detection calling image like `imageutils.GetE2EImage()`

**Comment by bclau** [comment]
> 
> Any good way to address this? because due to this we may end up having `agnhost:linux` and `agnhost:windows` kind of 2 images (harder to manage same functionality on both image, platform detection calling image like `imageutils.GetE2EImage()`

Well, the idea is to have ``agnhost`` as a manifest list, so we won't have to worry about the underlying host OS. I have worked on this previously, and the PR is still open at this moment: https://github.com/kubernetes/kubernetes/pull/76838 . So far, only the ``ss`` / ``netstat`` functionally will have to be addressed, and I don't have an definitive answer for that yet. Other than that, most of the differences between Linux and Windows are covered (e.g.: the Windows ``agnhost`` image is based on ``busybox``, so we have most of the Linux utilities as well).

I've already started working on centralizing the binaries / utilities used by different images into ``agnhost`` (``entrypoint-tester``, ``fakegitserver``, ``hostexec``, ``liveness``, ``logs-generator``, ``net``, ``netexec``, ``nettest``, ``no-snat-test``, ``no-snat-test-proxy``, ``porter``, ``port-forward-tester``, ``serve-hostname``). So far, I've noticed that most of these images are used in just a few places in the first place, so replacing them will be easy, with the exception of ``serve-hostname``, which is used quite often in tests. As for building the images, we'll have 2 Dockerfiles (one for Linux (which looks like the ``hostexec`` Dockerfile + ``agnhost`` binary), one for Windows (will only be based on the ``busybox`` image + ``agnhost`` binary)).

The ``agnhost`` binary right now has 7.8MB, (just the ``serve-hostname``, ``no-snat-test``, ``no-snat-test-proxy``, ``fakegitserver``, ``liveness`` binaries have ~29MB in total), so the overall image size will go down, it seems.

**Comment by bclau** [comment]
So, a small update regarding the image centralization. I've sent a WIP PR with all the changes here: https://github.com/kubernetes/kubernetes/pull/78161 . Ofc, it will have to be split into multiple PRs in order to be reviewable and mergeable, but some initial feedback would helpful.

Currently, I've built the ``claudiubelu/agnhost:1.1`` image to contain the following images (in addition with that it already had):

- fakegitserver
- liveness
- logs-generator
- net
- netexec
- nettest
- no-snat-test
- no-snat-test-proxy
- port-forward-tester
- serve-hostname
- webhook

The ``linux/amd64`` image has 40.3 MB in total (the ``gcr.io/kubernetes-e2e-test-images/webhook`` has 22MB).

Other images considered for centralization:

- audit-proxy (only used by a few tests)
- crd-conversion-webhook
- entrypoint-tester (only echoes the arguments, only used in a few tests)
- porter

What images probably won't be centralized:

- apparmor-loader (AFAIK, apparmor is Linux-specific, so it goes against the ``agnhost`` idea)
- metadata-concealment (the image only in a few tests requiring GCE, lower priority, IMO)
- nonewprivs (prints the EUID, which is a Linux-specific thing)

**Comment by spiffxp** [comment]
Versioning question that is tripping me up in PR reviews: I've seen bumps to `2.0` and `2.1`, is the intent for different groups of PRs to reference different versions, or move them all up at once?

**Comment by bclau** [comment]
Part 1 bumps to version 2.0, part 2 bumps to version 2.1.

Now, by mistake, Centralization part 2 merged before part 1, which is why it's a bit weird right now, but it should be better once part 1 merges.

**Comment by bclau** [comment]
Status update: Image centralization part 1 and part 2 merged. The tests using the images from part 1 have been updated to use the new ``agnhost:2.0`` image. For part 2, I'm waiting for ``agnhost:2.1`` to be built and published to gcr.io.

Added part 3 PR, which centralizes the images: audit-proxy, crd-conversion-webhook, entrypoint-tester, inclusterclient, iperf, porter, serve-hostname. It's still WIP ATM. Needs documentation updates, CLI helpers, testing. The new image version will be ``agnhost:2.2``.

Current ``agnhost`` binary size: 32 MB, but the total size of the binaries of all the centralized images is 171 MB, so, that looks better.

**Comment by johnSchnake** [comment]
/reopen

Will update with more details, but my last assessment was that there were a few more images that could/should be consolidated.

**Comment by johnSchnake** [comment]
The list of images still used by conformance runs according to my records is:
```
busybox
dnsutils
etcd
gb-frontend
gb-redisslave
ipc-utils
jessie-dnsutils
kitten
mounttest
mounttest-user
nautilus
nginx
pause
redis
resource-consumer
resource-consumer
sample-apiserver
test-webserver
volume/gluster
volume/nfs
```

The tests dealing with `nautilus`/`kitten`, which are both powered by the `test-webserver` are just basic web servers that report a given detail (json that says kitten or nautilus; I dont think the actual jpg gets checked ever). It seems like this could be rolled into our agnhost with some sort of 'static' endpoint that reports whatever we want.
 - [ ] Roll minimal behavior to cover kitten/nautilus into agnhost. Seems like the nature of the tests is just to see that it moved from imageX to imageY; seems like we could have some sort of switch at startup to expose some static json response. A single generic handler would work, but if we want a hardcoded `kitten` and `nautilus` we can; it will just lead to confusion again eventually about why we have those two as separate commands.
 - [ ] Remove need to have test-webserver at all (since the agnhost already is a little web server; I dont think this requires much if anything to deprecate)
 - [ ] mounttest image seems pretty simple to move to agnhost; it has a single go file and is otherwise made from scratch.
 - [ ] mounttest-user is the same as mounttest, but run with a given UID. Is it really important to have a completely separate user for that? Instead of specifically running with a user set in the security context? IMO it seems that would make it explicit what the difference between those two "images" are (just the UID)
 - [ ] What is the status of `pause`? It seems like it was rolled into agnhost but imageutils still has the separate pause image listed. Also has a helper defined there to specifically get that image (seems unnecessary and `GetE2EImage` should be used instead.

IMO if we iterate against those 6 images that is nearly as good as we can do for now. We _could_ roll the resource consumer in as well, but it is a little more complex than the others. The others only seem more complex in terms of their implementation or build requirements (OS, startup scripts, etc).

@bclau Thoughts on my final assessment about what we can/should integrate into that agnhost?

**Comment by bclau** [comment]
> 
> 
> The list of images still used by conformance runs according to my records is:
> 
> ```
> busybox
> dnsutils
> etcd
> gb-frontend
> gb-redisslave
> ipc-utils
> jessie-dnsutils
> kitten
> mounttest
> mounttest-user
> nautilus
> nginx
> pause
> redis
> resource-consumer
> resource-consumer
> sample-apiserver
> test-webserver
> volume/gluster
> volume/nfs
> ```
> 
> The tests dealing with `nautilus`/`kitten`, which are both powered by the `test-webserver` are just basic web servers that report a given detail (json that says kitten or nautilus; I dont think the actual jpg gets checked ever). It seems like this could be rolled into our agnhost with some sort of 'static' endpoint that reports whatever we want.
> 
>     * [ ]  Roll minimal behavior to cover kitten/nautilus into agnhost. Seems like the nature of the tests is just to see that it moved from imageX to imageY; seems like we could have some sort of switch at startup to expose some static json response. A single generic handler would work, but if we want a hardcoded `kitten` and `nautilus` we can; it will just lead to confusion again eventually about why we have those two as separate commands.

Right. IMO, making sure that the image actually changes on rolling updates, etc. is pretty important, So I would still keep them as separate images. Ofc, if we merge ``test-webserver`` into ``agnhost``, then ``kitten`` and ``nautilus`` will have to be based on ``agnhost``.

One thing that we could still do: update the tests to upgrade from ``agnhost`` to ``kitten`` or ``nautilus`` and remove the other. Thoughts?

> 
>     * [ ]  Remove need to have test-webserver at all (since the agnhost already is a little web server; I dont think this requires much if anything to deprecate)

Hm, not quite. ``test-webserver`` is basically a fileserver, function that we do not have in ``agnhost``. ``test-webserver`` is extremely small, we can easily include it in ``agnhost``.

> 
>     * [ ]  mounttest image seems pretty simple to move to agnhost; it has a single go file and is otherwise made from scratch.

Will have to add Windows support for it, but I already have the code for it (https://github.com/kubernetes/kubernetes/pull/77398/files#diff-05bcb687cc3bcbd82689804d2877f409R1).

> 
>     * [ ]  mounttest-user is the same as mounttest, but run with a given UID. Is it really important to have a completely separate user for that? Instead of specifically running with a user set in the security context? IMO it seems that would make it explicit what the difference between those two "images" are (just the UID)

Good point, from what I see, that image is only used in ``e2e/common/empty_dir.go`` tests. Just a small note: ``SecurityContext.RunAsUser`` does not work on Windows (but we merged this cycle: ``SecurityContext.WindowsContext.RunAsUserName``, so that can be used instead). Those tests focus on checking file permissions, so I don't think there's any reason why we can't go with your suggestion.

> 
>     * [ ]  What is the status of `pause`? It seems like it was rolled into agnhost but imageutils still has the separate pause image listed. Also has a helper defined there to specifically get that image (seems unnecessary and `GetE2EImage` should be used instead.
> 

Indeed, I've added the ``pause`` subcommand to ``agnhost``, since some tests required the pod to be Running, and it was useful when we had to run multiple exec commands into the pod. We can remove the ``Pause`` image and use ``agnhost`` instead.

> 
> IMO if we iterate against those 6 images that is nearly as good as we can do for now. We _could_ roll the resource consumer in as well, but it is a little more complex than the others. The others only seem more complex in terms of their implementation or build requirements (OS, startup scripts, etc).
> 

The ``resource-consumer`` and ``resource-consumer-controller`` seem doable to me, and it was on my shortlist. Ofc, I will also have to add Windows support for it, but the work for that was already done (https://github.com/kubernetes/kubernetes/pull/77398/files#diff-9aa5b01e9f770e378f1f542e22a480c6R1)

> @bclau Thoughts on my final assessment about what we can/should integrate into that agnhost?

I've added my comments above. I'm also thinking about other images as well.

``busybox``- well, I don't think we need it anymore, since ``agnhost`` is based on``alpine:3.6``. It should contain anything that ``busybox`` does. We could try simply replacing it with ``agnhost``.

``dnsutils``

- It basically contains bind-tools and CoreDNS. I have a PR which adds bind-tools to ``agnhost`` (https://github.com/kubernetes/kubernetes/pull/81008). We could also add CoreDNS to ``agnhost`` and be done with the ``dnsutils`` image.
- One extra reason to do this: a PR changed some DNS tests to use ``getent hosts`` instead of ``hostname -i``, which doesn't work on Windows (``getent`` does not exist on Windows, and the ``/etc/hosts`` file is not managed by Kubernetes on Windows, for now). I couldn't find a platform-agnostic command which would get the IPv4 / IPv6 required by the test, so I was thinking of adding an ``agnhost`` subcommand for it.
- But, there is still ``jessie-dnsutils``, and we can't give up on it yet... soo, we could have instead ``jessie-agnhost`` which will be based on the jessie image? What are your opinions?

**Comment by johnSchnake** [comment]
> making sure that the image actually changes on rolling updates, etc. is pretty important, So I would still keep them as separate images

There are a few tests like this it seems but as far as I know that is also the reason for having "nginx" and "nginxNew". We can probably decide on just one approach to validate the "ensure image gets updated" logic get rid of `nginxNew` then.

> busybox- well, I don't think we need it anymore, since agnhost is based onalpine:3.6. It should contain anything that busybox does. We could try simply replacing it with agnhost.

That would be great; I didn't even consider that really since I was just thinking about what could be "rolled into" agnhost. You're probably right that agnhost would probably work in all those busybox cases.

 - WRT the dnsutils, you are more ambitious than me :) That is definitely starting to get more complex so if you feel up to getting that completely gone then that is fine with me. I was just ballparking that the effort was significantly higher than many of the other images.

 - WRT `jessie-xyz` I don't feel informed enough to comment on that. I think you've had a few indepth conversations about why we have that and what makes it special though. I'd defer to you in general.

- WRT `resource-consumer[-controller]` I thought those were medium effort compared to the others. If they were on your shortlist I'd 👍 them being merged in as well. I didnt see anything that would _not_ make me want them in.

So, in summary it sounds like we have:
 - [ ] Update tests to use `agnhost` instead of `pause`
 - [ ] Update tests to use `agnhost` instead of `busybox`
 - [ ] Integrate test-webserver (a file server) into `agnhost` and base `kitten/nautilus` on `agnhost`
 - [ ] Evaluate `nginxNew` tests and probably change them to use kitten/nautilus images to ensure image-updating behavior
 - [ ] Move `mounttest` into `agnhost`
 - [ ] Update tests that use `mounttest-user` to simply use `mounttest` and update the podspecs to specify a user. I think this item is open for debate though.
 - [ ] Complete work to get dns functionality into the `agnhost` image which means bind-tools and coreDNS. The first half of that is pending merge already.
 - [ ] Continue discussion about `jessie-dnsutils`

Is that accurate? I can definitely help with some of those items. It would be easiest for me to tackle the updating of tests since you have the experience at this point of iterating on and getting new images built/tested.

**Comment by bclau** [comment]
> So, in summary it sounds like we have:
> 
>     * [ ]  Integrate test-webserver (a file server) into `agnhost` and base `kitten/nautilus` on `agnhost`
> 
>     * [ ]  Move `mounttest` into `agnhost`
> 
>     * [ ]  Complete work to get dns functionality into the `agnhost` image which means bind-tools and coreDNS. The first half of that is pending merge already.

I've started working on those items + resource-consumer and resource-consumer-controller. Will send a WIP PR soon, and I'm expecting to finish it by tomorrow.

> 
> Is that accurate? I can definitely help with some of those items. It would be easiest for me to tackle the updating of tests since you have the experience at this point of iterating on and getting new images built/tested.

Sure, that would be greatly appreciated. You can simply go ahead with the following items:

>     * [ ]  Update tests to use `agnhost` instead of `pause`
> 
>     * [ ]  Update tests to use `agnhost` instead of `busybox`
> 
>     * [ ]  Update tests that use `mounttest-user` to simply use `mounttest` and update the podspecs to specify a user. I think this item is open for debate though.
> 

Those items do not require a new ``agnhost`` image to be built, it should work with the current version too. Just a fair warning, the ``busybox`` image seems to be used quite extensively.

**Comment by aojea** [comment]
@johnSchnake @bclau , I´d like to sync with you because I´m touching some tests so they are able to run on both IPv4 and IPv6 without any code change. 
One of the changes I`ve introduced is switching the nginx images by httpd images because the later listen on both IPv4 and IPv6 by default.

**Comment by bclau** [comment]
> 
> 
> @johnSchnake @bclau , I´d like to sync with you because I´m touching some tests so they are able to run on both IPv4 and IPv6 without any code change.
> One of the changes I`ve introduced is switching the nginx images by httpd images because the later listen on both IPv4 and IPv6 by default.

Can you list those tests so I can take a look as well?

Additionally, with the latest Centralization PR, I've included ``test-webserver`` into agnhost as you've previously wanted.

**Comment by aojea** [comment]
>
> Can you list those tests so I can take a look as well?

Sure, let me know if this link works https://github.com/pulls?utf8=%E2%9C%93&q=is%3Apr+author%3Aaojea+archived%3Afalse+is%3Amerged+e2e+IPv6+user%3Akubernetes+

Typically, is only using net.JoinHosPort to create IPv6 URLs, but sometimes images doesn't work for both IPv4 and IPv6. IIRC I've replaced ngingx by httpd and migrated the redis image




> 
> Additionally, with the latest Centralization PR, I've included `test-webserver` into agnhost as you've previously wanted.

That's great, greping for nginx and httpd in the e2e folder there are a lot of matches, I think that all those images can be centrilized in agnost

**Comment by xmudrii** [comment]
As this seems to have no activity and the code freeze started, I'm moving it to the next milestone.

/milestone v1.17

**Comment by smourapina** [comment]
@johnSchnake Bug triage for 1.17 here. This issue has been open for a significant amount of time and since it is tagged for milestone 1.17, we want to let you know that the 1.17 code freeze is coming in less than one month on November 18th. Will this issue be resolved before then?

**Comment by johnSchnake** [comment]
I think that we have followed this issue to a reasonable conclusion; there are 3 pending PRs up that will further image consolidation and fixup some tests.

I will take a look at the "and improve documentation" part of this today and see if there is some clear work that needs done on that front.

I don't think anything beyond those tickets needs done and this is just an umbrella ticket for those really. 🤞 that it can all get done for the code freeze but https://github.com/kubernetes/kubernetes/pull/81170 seems to be waiting on some Q&A that I'm not equipped to comment on.

**Comment by smourapina** [comment]
@bclau @johnSchnake Friendly reminder that code freeze is rapidly approaching and will take place in November 14th. Is this issue still on track for 1.17? Otherwise we will need to move it to 1.18 milestone.

**Comment by markyjackson-taulia** [comment]
Hello, @bclau @johnSchnake
I wanted to check in on this issue to ensure it was still on track for the v1.18 release cycle.

**Comment by claudiubelu** [comment]
It is. We've merged the Image Centralization part 4 PR recently, a few fixes need to merge, and then this will have to also merge: https://github.com/kubernetes/kubernetes/pull/81226

**Comment by markyjackson-taulia** [comment]
Hello, @claudiubelu gentle ping here in regards to this issue.
Are we close to completing this so it will make it in the v1.18 release? Code freeze is coming soon

**Comment by markyjackson-taulia** [comment]
Hello, @claudiubelu,
Are we close to completing this? Code Freeze is coming soon
![image](https://user-images.githubusercontent.com/16655670/74668939-15edc800-515b-11ea-8950-2123cf6860ea.png)

**Comment by claudiubelu** [comment]
> 
> 
> Hello, @claudiubelu,
> Are we close to completing this? Code Freeze is coming soon
> ![image](https://user-images.githubusercontent.com/16655670/74668939-15edc800-515b-11ea-8950-2123cf6860ea.png)

From my part, we need this part in: https://github.com/kubernetes/kubernetes/pull/81226

**Comment by markyjackson-taulia** [comment]
Hello, @claudiubelu 

Code freeze is 3/5, do you think we are going to complete #88248 and #88249 in time?

**Comment by markyjackson-taulia** [comment]
@claudiubelu do you think this is going to make the v1.18 code freeze date? 3/5 is rapidly approaching. If you don’t think so, let me know so I can move this in to the next milestone

**Comment by claudiubelu** [comment]
^ The above PRs have received some comments and one lgtm. They're not very big, nor complicated.

**Comment by markyjackson-taulia** [comment]
@claudiubelu we are 24 hours out from code freeze. Our options are:
1: move to the the next milestone
2: request an exception 

Given the 2 items that still are in progress above what is your confidence level these will be complete without an exception?

**Comment by markyjackson-taulia** [comment]
I have sent a slack dm and will wait till mid day pst time for a response.
